### ISSUE TITLE
ktlint --stdin dies with StackOverflowError on syntax error

### ISSUE THREAD

**Issue by derat** [pinterest/ktlint#2379]
Whenever I pass input with a syntax error to `ktlint --stdin`, it hangs for a while and then crashes.

## Expected Behavior

ktlint should report a single syntax error and exit.

## Observed Behavior

ktlint repeatedly logs a syntax error and a message about trying to read the input as a Kotlin Script  for a (potentially long) while before dying with a `java.lang.StackOverflowError`.

## Steps to Reproduce

```
$ echo -e 'package org.example.test\nfun foo() = ' | ktlint --stdin
18:23:07.003 [main] INFO com.pinterest.ktlint.cli.internal.KtlintCommandLine - Enable default patterns [**/*.kt, **/*.kts]
18:23:07.377 [main] WARN com.pinterest.ktlint.cli.internal.KtlintCommandLine - Can not parse input from <stdin> as Kotlin, due to error below:
    Not a valid Kotlin file (2:12 expecting an expression)
Now, trying to read the input as Kotlin Script.
18:23:07.398 [main] WARN com.pinterest.ktlint.cli.internal.KtlintCommandLine - Can not parse input from <stdin> as Kotlin, due to error below:
    Not a valid Kotlin file (2:12 expecting an expression)
Now, trying to read the input as Kotlin Script.
18:23:07.403 [main] WARN com.pinterest.ktlint.cli.internal.KtlintCommandLine - Can not parse input from <stdin> as Kotlin, due to error below:
    Not a valid Kotlin file (2:12 expecting an expression)
Now, trying to read the input as Kotlin Script.
...
Exception in thread "main" java.lang.StackOverflowError                                                                                                      
        at java.base/java.lang.Exception.<init>(Exception.java:67)
...
        at io.github.oshai.kotlinlogging.slf4j.internal.LocationAwareKLogger.warn(LocationAwareKLogger.kt:190)
        at com.pinterest.ktlint.cli.internal.KtlintCommandLine.lint(KtlintCommandLine.kt:569)
        at com.pinterest.ktlint.cli.internal.KtlintCommandLine.lint(KtlintCommandLine.kt:576)                                                               
        at com.pinterest.ktlint.cli.internal.KtlintCommandLine.lint(KtlintCommandLine.kt:576)
        at com.pinterest.ktlint.cli.internal.KtlintCommandLine.lint(KtlintCommandLine.kt:576)
        at com.pinterest.ktlint.cli.internal.KtlintCommandLine.lint(KtlintCommandLine.kt:576)
...
```

## Your Environment

* Version of ktlint used: 1.0.1, 1.0.0
* Relevant parts of the `.editorconfig` settings: no `.editorconfig` needed
* Name and version of integration used: N/A
* Version of Gradle used (if applicable): N/A
* Operating System and version: Debian 11.8 x86_64

## Notes

This code should probably avoid the recursive call to `lint()` when it's already trying to read the input as a Kotlin Script:

https://github.com/pinterest/ktlint/blob/456d9281baab6ad1660f0f036e1fa7a1529a02a7/ktlint-cli/src/main/kotlin/com/pinterest/ktlint/cli/internal/KtlintCommandLine.kt#L588-L606
